docs/src/keymap-upgrade/index.ts

```diff
@@ -0,0 +1,59 @@
+import { createParser } from "./parser";
+import { applyEdits } from "./textedit";
+
+import { upgradeBehaviors } from "./behaviors";
+import { upgradeHeaders } from "./headers";
+import { upgradeKeycodes } from "./keycodes";
+import { upgradeProperties } from "./properties";
+
+export { initParser } from "./parser";
+
+const upgradeFunctions = [
+  upgradeBehaviors,
+  upgradeHeaders,
+  upgradeKeycodes,
+  upgradeProperties,
+];
+
+export function upgradeKeymap(text: string) {
+  const parser = createParser();
+  const tree = parser.parse(text);
+
+  const edits = upgradeFunctions.map((f) => f(tree)).flat();
+
+  return applyEdits(text, edits);
+}
+
+export function rangesToLineNumbers(
+  text: string,
+  changedRanges: [number, number][]
+): string {
+  const lineBreaks = getLineBreakPositions(text);
+
+  const changedLines = changedRanges.map(([start, end]) => {
+    const startLine = positionToLineNumber(start, lineBreaks);
+    const endLine = positionToLineNumber(end, lineBreaks);
+
+    return startLine === endLine ? `${startLine}` : `${startLine}-${endLine}`;
+  });
+
+  return `{${changedLines.join(",")}}`;
+}
+
+function getLineBreakPositions(text: string) {
```

Comment:
Can this be simplified by something like below?
```ts
Array.from(text).map((c, i) => c === '\n' ? i : -1).filter(i => i !== -1)
```
(Apparently `Array.map` [assigns the index to the second arg](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#callbackfn) for the callback)

Reply:
It could be written more simply like that, but this implementation is [much faster](https://jsperf.app/yatiqa/3) on account of scanning through the original string without modifying it instead of turning it into an array of one-character strings.

Reply:
Interesting, that's indeed a very large difference. Thanks for the info.